File: solutions/cpp/0872-leaf-similar-tree.cpp
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> leaves1;
        vector<int> leaves2;
        dfs(root1, leaves1);
        dfs(root2, leaves2);
        return leaves1 == leaves2;
    }

private:
    void dfs(TreeNode* node, vector<int>& leaves) {
        if (node == nullptr) return;
        if (node->left == nullptr && node->right == nullptr) {
            leaves.push_back(node->val);
        }
        dfs(node->left, leaves);
        dfs(node->right, leaves);
    }
};

File: solutions/javascript/0872-leaf-similar-tree.js
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    const dfs = (node, leaves) => {
    if (node === null) return;
    if (!node.left && !node.right) {
        leaves.push(node.val);
    }
    dfs(node.left, leaves);
    dfs(node.right, leaves);
    }

    const leaves1 = [], leaves2 = [];
    dfs(root1, leaves1);
    dfs(root2, leaves2);
    return JSON.stringify(leaves1) === JSON.stringify(leaves2);
};
